MapboxVision/Services/VideoSource/CameraVideoSource.swift

```diff
@@ -0,0 +1,162 @@
+        guard let pixelBuffer = sampleBuffer.pixelBuffer else { return nil }
+        
+        let width = pixelBuffer.width
+        let height = pixelBuffer.height
+        
+        let focalPixelX: Float?
+        let focalPixelY: Float?
+        
+        if let attachment = CMGetAttachment(sampleBuffer, kCMSampleBufferAttachmentKey_CameraIntrinsicMatrix, nil) as? Data {
+            let matrix: matrix_float3x3 = attachment.withUnsafeBytes { $0.pointee }
+            focalPixelX = matrix[0,0]
+            focalPixelY = matrix[1,1]
+        } else {
+            focalPixelX = formatFieldOfView
+            focalPixelY = formatFieldOfView
+        }
+        
+        return CameraParameters(width: width, height: height, focalXPixels: focalPixelX, focalYPixels: focalPixelY)
+    }
+    
+    private func notify(closure: (VideoSourceObserver) -> Void) {
+        observations.forEach { (id, observation) in
+            guard let observer = observation.observer else {
+                observations.removeValue(forKey: id)
+                return
+            }
+            closure(observer)
+        }
+    }
+    
+    private var formatFieldOfView: Float? {
+        guard let fov = camera?.activeFormat.videoFieldOfView else { return nil }
+        return fov > 0 ? fov : nil
+    }
+    
+    // MARK: - Observations
+    
+    @objc private func orientationChanged() {
+        dataOutput?.connection(with: .video)?.set(deviceOrientation: UIDevice.current.orientation)
+    }
+}
+
+extension CameraVideoSource: VideoSource {
+    public func add(observer: VideoSourceObserver) {
+        let id = ObjectIdentifier(observer)
+        observations[id] = Observation(observer: observer)
+    }
+    
+    public func remove(observer: VideoSourceObserver) {
+        let id = ObjectIdentifier(observer)
+        observations.removeValue(forKey: id)
+    }
+    
+    open var isExternal: Bool {
+        return false
+    }
+}
+
+extension CameraVideoSource: AVCaptureVideoDataOutputSampleBufferDelegate {
+    public func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
+        notify { (observer) in
+            let sample = VideoSample(buffer: sampleBuffer, format: .bgra)
+            observer.videoSource(self, didOutput: sample)
+            
+            if let cameraParameters = getCameraParameters(sampleBuffer: sampleBuffer) {
+                observer.videoSource(self, didOutput: cameraParameters)
+            }
+        }
+    }
+    
+    public func captureOutput(_ output: AVCaptureOutput, didDrop sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
+        var mode: CMAttachmentMode = 0
+        guard let reason = CMGetAttachment(sampleBuffer, kCMSampleBufferAttachmentKey_DroppedFrameReason, &mode) else { return }
+        print("Sample buffer was dropped. Reason: \(reason)")
```

Comment:
Is print needed or added just for debug purposes?

Reply:
`print` is our current logging system.